### ISSUE TITLE
Under high load heartbeats can have concurrency issues

### ISSUE THREAD

**Issue by mikeminutillo** [Particular/ServiceControl#639]
When ServiceControl receives a heartbeat message it loads a document from it's persistence, updates it and saves it back again. Under high load two incoming heartbeats from the same endpoint can be worked on concurrently causing a concurrency exception. Even when one of these gets a concurrency exception and rolls back, if it is earlier than the one successfully processed then ServiceControl will discard the older one when it tries to re-process it. This further clogs up the already congested ServiceControl input queue.

I would propose the following solution:

1. Keep all heartbeat information in an in-memory concurrent dictionary keyed on endpoint id.
2. Serve all heartbeat queries from this dictionary.
3. Persist the dictionary to Raven on a predictable schedule (say every 2 minutes) rather than on every heartbeat.
4. When a user toggles heartbeats on / off in Pulse this will need to persist immediately. The same when endpoints transition from Dead to Heartbeating and back again.

Pros: 
- Saving heartbeats will (almost) never cause concurrency exceptions. Toggling heartbeats on/off and state transitions may still cause this.

Cons: 
- As the data is in memory, we are restricted to a single instance. ServiceControl is already limited to a single instance by the failed message recovery mechanism. 
- If ServiceControl shuts down unexpectedly between modifying and saving the data we may lose the latest heartbeats. If the endpoint is still heartbeating when ServiceControl comes back online we will update anyway. If the endpoint is not heartbeating when ServiceControl comes back online then the difference of a few heartbeats as to when it went down is unlikely to be very relevant. 

NOTES:
- We already DO keep an in-memory cache of heartbeats and we use it to determine when an endpoint has stopped heartbeating. 
- When we detect an endpoint that's not heartbeating we send a message to ourselves. This message can be enqueued behind a heartbeat message from that endpoint so that when we get to processing it, the heartbeat has already been restored. If heartbeats are meant to be low latency we should probably handle this synchronously as soon as it is discovered if possible.

**Comment by mikeminutillo** [comment]
@MarcinHoppe this is the issue that @andreasohlund was talking about.

@mauroservienti @sfarmar @ramonsmits as you've had recent dealings with customers about SC I'd appreciate your input here 

Heartbeats have another issue as well which is that they can bank up in the ServiceControl queue if the transport does not handle TTBR well. https://github.com/Particular/NServiceBus.SqlServer/issues/158

**Comment by mauroservienti** [comment]
I observed this happening last week, the failure rate was not really high but the log was filled with concurrency errors due to heartbeats.

Question: why are we updating a single document when dealing with incoming heartbeats instead of relying on an asyn index to gather heartbeats stats? If this is the case.

**Comment by mikeminutillo** [comment]
@mauroservienti I wonder then how much this is a real problem and how much it is just log noise. Would simply upping the log level to DEBUG resolve a lot of this?

> Question: why are we updating a single document when dealing with incoming heartbeats instead of relying on an asyn index to gather heartbeats stats? If this is the case.

I suspect due to the low latency nature of the feature. If we decide that an endpoint is down then we need to notify users immediately. We could still do that with an in memory solution and persist at state transitions though. 

i.e. When we detect an endpoint has stopped heartbeating, update it's status in memory, persist to Raven and (if monitored) raise an event. When we detect an endpoint has started beating, update it's status, persist to Raven and (if monitored) raise an event. When a user tells us to start/stop monitoring an endpoint, update it in memory and persist in Raven. Every 2 minutes, persist in Raven. 

Those persistence steps need to be serialized or we'll get concurrency exceptions again.

**Comment by mikeminutillo** [comment]
Realistically the every 2 minutes thing is only needed for "Last Seen". How much value do we get from that. Is it shown in the UI? If we were to lose that during a restart, we'd either immediately be getting heartbeats from the live endpoint or we'd be getting no heartbeats from a dead endpoint. If we are down for 20 minutes, does it matter when we last saw an endpoint before this session?

**Comment by andreasohlund** [comment]
Should this be labeled as a bug? Should we patch this?

**Comment by johnsimons** [comment]
Yes we are patching it

**Comment by andreasohlund** [comment]
>Yes we are patching it

Can assign a milestone to this one? (https://github.com/Particular/PlatformDevelopment/issues/596)

**Comment by johnsimons** [comment]
Yes, once we know when is going out, dependent on #640 being complete

**Comment by mikeminutillo** [comment]
@andreas the SC release process is quite a bit less strict than the Core one because we have no need to support older versions. As a result when you say "are we patching this" and John says yes, it just means that it will go into a release when it is ready.

**Comment by andreasohlund** [comment]
Good point, so how do we know that this is a prio? 

`vnext` milestone? label? the "maintainers just know"? other idea?

**Comment by mauroservienti** [comment]
> "maintainers just know"

Dangerous, IMO.
Doesn't the waffle board sort order already define this? (Just asking, maybe the sort order us not even persisted)

**Comment by johnsimons** [comment]
Fixed in https://github.com/Particular/ServiceControl/pull/664
